### PR TITLE
Use providerSpec instead of providerConfig

### LINKED ISSUE #416: Use machine.spec.providerSpec instead of machine.spec.providerConfig
The cluster-api project merged a pull that moved `.spec.providerConfig` to `.spec.ProviderSpec`: https://github.com/kubernetes-sigs/cluster-api/pull/548

In order to be able to update the cluster-api dependency, we must write a migration for this and change our code accordingly

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

* Adds a migration to migrate the .spec.prociderConfig field of existing machines to .spec.providerSpec
* Changes the code to use .spec.providerSpec to .spec.ProviderConfig
* Updates the cluster-api dependency to [45f1c93260140936c610e56575d7505ba3d52444](https://github.com/kubernetes-sigs/cluster-api/tree/45f1c93260140936c610e56575d7505ba3d52444)

This allows us to update the cluster-api dependency

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #416

**Special notes for your reviewer**:

```release-note
Rename `machine.spec.providerConfig` to `machine.spec.providerSpec`
```

/assign @mrIncompetent

### CODE REVIEW

File: pkg/admission/machines.go [hunk 1]
```diff
@@ -39,13 +44,18 @@ func (ad *admissionData) mutateMachines(ar admissionv1beta1.AdmissionReview) (*a
 		if oldMachine.Spec.Name != machine.Spec.Name && machine.Spec.Name == machine.Name {
 			oldMachine.Spec.Name = machine.Spec.Name
 		}
-		// Allow mutation when oldMachine has Initializers on it
-		if oldMachine.Initializers == nil || len(oldMachine.Initializers.Pending) == 0 {
+		// Allow mutation when:
+		// * oldMachine has Initializers on it
+		// * machine has the `MigrationBypassSpecNoModificationRequirementAnnotation` annotation (used for type migration)
+		_, hasBypassSpecNoModificationRequirementAnnotation := machine.Annotations[BypassSpecNoModificationRequirementAnnotation]
```
Comment (alvaroaleman) [inline pkg/admission/machines.go]: Not sure if this is the most elegant approach. Definitely open for ideas here

File: pkg/userdata/ubuntu/userdata.go [hunk 2]
```diff
@@ -62,7 +62,7 @@ func (p Provider) UserData(
 		return "", fmt.Errorf("failed to get cloud config: %v", err)
 	}
 
-	pconfig, err := providerconfig.GetConfig(spec.ProviderConfig)
+	pconfig, err := providerconfig.GetConfig(spec.ProviderSpec)
 	if err != nil {
 		return "", fmt.Errorf("failed to get provider config: %v", err)
```
Comment (mrIncompetent) [inline pkg/userdata/ubuntu/userdata.go]: Should be replaced with `failed to get provider spec`

Would be great if you could grep the whole code for `provider config` and replace it

File: pkg/admission/machinedeployments_validation.go [hunk 3]
```diff
@@ -90,47 +90,6 @@ func getIntOrPercent(s *intstr.IntOrString, roundUp bool) (int, error) {
...
-		if obj.Spec.Strategy.RollingUpdate.MaxUnavailable == nil {
-			x := intstr.FromInt(0)
-			obj.Spec.Strategy.RollingUpdate.MaxUnavailable = &x
-		}
-	}
-	if len(obj.Namespace) == 0 {
-		obj.Namespace = metav1.NamespaceDefault
-	}
+func machineDeploymentDefaultingFunction(md *v1alpha1.MachineDeployment) {
+	v1alpha1.PopulateDefaultsMachineDeployment(md)
```
Comment (mrIncompetent) [inline pkg/admission/machinedeployments_validation.go]: :tada:

File: pkg/admission/machines.go [hunk 4]
```diff
@@ -17,6 +17,11 @@ import (
 	clusterv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
 )
 
+// BypassSpecNoModificationRequirementAnnotation is used to bypass the "no machine.spec modification" allowed
+// restriction from the webhook in order to change the spec in some special cases, e.G. for the migration of
+// the `providerConfig` field to `providerSpec`
+const BypassSpecNoModificationRequirementAnnotation = "kubermatic.io/bypass-no-spec-mutation-requrirement"
```
Comment (mrIncompetent) [inline pkg/admission/machines.go]: `requrirement` -> `requirement`

File: pkg/admission/machines.go (hunk 1 above)
Comment (mrIncompetent) [reply pkg/admission/machines.go]: I would prefer checking the value to be more precise:
```
bypassValidationForMigration := machine.Annotations[BypassSpecNoModificationRequirementAnnotation] == "true"
```

But no hard opinion here - i dont think the `exists-in-map` approach is bad

File: pkg/apis/cluster/v1alpha1/conversions/providerconfig_to_providerspec_test.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+			}
+		}
+
+		resultFixtureContent, err := ioutil.ReadFile(resultFixturePath)
+		if err != nil {
+			t.Errorf("failed to read result fixture for machine %s: %v", convertedMachine.Name, err)
+			continue
+		}
+
+		if string(convertedMachineYamlBytes) != string(resultFixtureContent) {
```
Comment (mrIncompetent) [inline pkg/apis/cluster/v1alpha1/conversions/providerconfig_to_providerspec_test.go]: Can we rather use https://github.com/kubermatic/machine-controller/blob/master/pkg/test/helper.go#L11 ?
It offers the update functionality + a git diff

**Comment by alvaroaleman** [comment]
@mrIncompetent Updated, PTAL

**Pull request merged**